src/finch/finchassembly/nodes.py

```diff
@@ -0,0 +1,373 @@
+    def is_expr():
+        """Determines if the node is an expression."""
+        return False
+
+    @staticmethod
+    def is_stateful():
+        """Determines if the node is stateful."""
+        return True
+
+    def children(self):
+        """Returns the children of the node."""
+        return [self.buffer, self.var, self.body]
+
+    @classmethod
+    def make_term(cls, head, *args):
+        return head(*args)
+
+@dataclass(eq=True, frozen=True)
+class WhileLoop(AssemblyNode):
+    """
+    Represents a while loop that executes as long as the condition is true.
+
+    Attributes:
+        condition: The condition to evaluate for the loop to continue.
+        body: The body of the loop to execute.
+    """
+
+    condition: AssemblyNode
+    body: AssemblyNode
+
+    @staticmethod
+    def is_expr():
+        """Determines if the node is an expression."""
+        return False
+
+    @staticmethod
+    def is_stateful():
+        """Determines if the node is stateful."""
+        return True
+
+    def children(self):
+        """Returns the children of the node."""
+        return [self.condition, self.body]
+
+    @classmethod
+    def make_term(cls, head, *args):
+        return head(*args)
+
+
+@dataclass(eq=True, frozen=True)
+class Assignment(AssemblyNode):
+    """
+    Represents a logical AST statement that evaluates `rhs`, binding the result to `lhs`.
+
+    Attributes:
+        type: The type of the new binding.
+        lhs: The left-hand side of the binding.
+        rhs: The right-hand side to evaluate.
+    """
+
+    type: AssemblyNode
+    lhs: AssemblyNode
+    rhs: AssemblyNode
+
+    @staticmethod
+    def is_expr():
+        """Determines if the node is an expression."""
+        return True
+
+    @staticmethod
+    def is_stateful():
+        """Determines if the node is stateful."""
+        return True
+
+    def children(self):
+        """Returns the children of the node."""
+        return [self.type, self.lhs, self.rhs]
+
+
+@dataclass(eq=True, frozen=True)
+class Return(AssemblyNode):
+    """
+    Represents a return statement that returns `arg` from the current function.
+    Halts execution of the function body.
+
+    Attributes:
+        arg: The argument to return.
+    """
+
+    args: tuple[AssemblyNode]
+
+    @staticmethod
+    def is_expr():
+        """Determines if the node is an expression."""
+        return True
+
+    @staticmethod
+    def is_stateful():
+        """Determines if the node is stateful."""
+        return True
+
+    def children(self):
+        """Returns the children of the node."""
+        return [self.arg]
+
+    @classmethod
+    def make_term(cls, head, *args):
+        return head(args)
+
+
+@dataclass(eq=True, frozen=True)
+class Block(AssemblyNode):
+    """
+    Represents a statement that executes a sequence of statements `bodies...`.
+
+    Attributes:
+        bodies: The sequence of statements to execute.
+    """
+
+    bodies: tuple[AssemblyNode] = ()
```

Comment:
In general, `tuple[T]` represents a type with a single element of type `T`. If we want an arbitrary number of elements all of type `T`, we should use `tuple[T, ...]`.

Note that `list` is variable-sized by default so for that, `list[T]` suffices.

Reply:
```suggestion
    bodies: tuple[AssemblyNode, ...] = ()
```